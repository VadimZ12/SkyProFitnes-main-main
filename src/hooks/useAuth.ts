import { useState, useEffect } from 'react';
import { User } from '../types/interfaces';
import { auth, database } from '../config/firebase';
import { onAuthStateChanged, signOut, updateProfile as updateFirebaseProfile, User as FirebaseUser } from 'firebase/auth';
import { ref, get, set } from "firebase/database";
import { registerUser, loginUser, resetPassword, changePassword } from '../api/auth';

const INACTIVITY_TIMEOUT = 3 * 60 * 1000; // 3 минуты в миллисекундах

export const useAuth = () => {
  const [user, setUser] = useState<User | null>(null);
  const [loading, setLoading] = useState(true);

  const updateLastActivity = () => {
    localStorage.setItem('lastActivity', Date.now().toString());
  };

  const checkInactivity = () => {
    const lastActivity = localStorage.getItem('lastActivity');
    if (lastActivity) {
      const inactiveTime = Date.now() - parseInt(lastActivity);
      if (inactiveTime > INACTIVITY_TIMEOUT) {
        logout();
      }
    }
  };

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, async (firebaseUser: FirebaseUser | null) => {
      try {
        if (firebaseUser) {
          checkInactivity();
          updateLastActivity();
          const userRef = ref(database, `users/${firebaseUser.uid}`);
          const snapshot = await get(userRef);
          if (snapshot.exists()) {
            const userData = snapshot.val();
            setUser({
              uid: firebaseUser.uid,
              email: firebaseUser.email!,
              displayName: userData.displayName || firebaseUser.displayName,
              customDisplayName: userData.customDisplayName,
            });
          } else {
            setUser({
              uid: firebaseUser.uid,
              email: firebaseUser.email!,
              displayName: firebaseUser.displayName || undefined,
            });
          }
        } else {
          setUser(null);
        }
      } catch (error) {
        console.error('Error in auth state change:', error);
        setUser(null);
      } finally {
        setLoading(false);
      }
    });

    window.addEventListener('mousemove', updateLastActivity);
    window.addEventListener('keypress', updateLastActivity);

    return () => {
      unsubscribe();
      window.removeEventListener('mousemove', updateLastActivity);
      window.removeEventListener('keypress', updateLastActivity);
    };
  }, []);

  const register = async (email: string, password: string) => {
    try {
      const user = await registerUser(email, password);
      updateLastActivity();
      setUser({
        uid: user.uid,
        email: user.email!,
        displayName: user.displayName || undefined,
      });
    } catch (error) {
      console.error('Error in register:', error);
      throw error;
    }
  };

  const login = async (email: string, password: string) => {
    try {
      const user = await loginUser(email, password);
      updateLastActivity();
      setUser({
        uid: user.uid,
        email: user.email!,
        displayName: user.displayName || undefined,
      });
    } catch (error) {
      console.error('Error in login:', error);
      throw error;
    }
  };

  const logout = async () => {
    try {
      await signOut(auth);
      localStorage.removeItem('lastActivity');
      setUser(null);
    } catch (error) {
      console.error('Error in logout:', error);
      throw error;
    }
  };

  const resetUserPassword = async (email: string) => {
    try {
      await resetPassword(email);
    } catch (error) {
      console.error('Error in resetPassword:', error);
      throw error;
    }
  };

  const changeUserPassword = async (newPassword: string) => {
    try {
      await changePassword(newPassword);
    } catch (error) {
      console.error('Error in changePassword:', error);
      throw error;
    }
  };

  const updateUser = async (updatedUser: Partial<User>) => {
    if (auth.currentUser) {
      try {
        const userRef = ref(database, `users/${auth.currentUser.uid}`);
        await updateFirebaseProfile(auth.currentUser, {
          displayName: updatedUser.displayName,
        });
        await set(userRef, updatedUser);
        setUser(currentUser => {
          if (currentUser) {
            return { ...currentUser, ...updatedUser };
          }
          return currentUser;
        });
      } catch (error) {
        console.error('Error updating user profile:', error);
        throw error;
      }
    }
  };

  return {
    user,
    loading,
    register,
    login,
    logout,
    resetUserPassword,
    changeUserPassword,
    updateUser,
  };
};