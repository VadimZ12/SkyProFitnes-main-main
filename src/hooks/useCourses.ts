import { useState, useEffect, useCallback } from 'react';
import { Course, Workout, User } from '../types/interfaces';
import { getAllCourses, addCourseToUser, getUserCourses as getUserCoursesAPI, getCourseById, getWorkoutById } from '../api/courses';
import { getUserProfile as getUserProfileFromAPI, updateUserProfile } from '../api/user';
import { useAuth } from './useAuth';
import { database } from '../config/firebase';
import { ref, get, remove } from "firebase/database";

const CACHE_EXPIRATION_TIME = 8 * 60 * 60 * 1000; // 8 часов в миллисекундах
interface CachedData<T> {
  data: T;
  timestamp: number;
}

const getCachedData = <T>(key: string): T | null => {
  try {
    const cachedItem = localStorage.getItem(key);
    if (cachedItem) {
      const { data, timestamp }: CachedData<T> = JSON.parse(cachedItem);
      if (Date.now() - timestamp < CACHE_EXPIRATION_TIME) {
        return data;
      }
    }
  } catch (error) {
    console.error('Error reading from localStorage:', error);
  }
  return null;
};

const setCachedData = <T>(key: string, data: T) => {
  try {
    const cacheItem: CachedData<T> = {
      data,
      timestamp: Date.now(),
    };
    localStorage.setItem(key, JSON.stringify(cacheItem));
  } catch (error) {
    console.error('Error writing to localStorage:', error);
  }
};

export const useCourses = () => {
  const [courses, setCourses] = useState<Course[]>([]);
  const [userCourses, setUserCourses] = useState<Course[]>([]);
  const [loading, setLoading] = useState(true);
  const { user: authUser } = useAuth();

  const fetchCourses = useCallback(async () => {
    setLoading(true);
    try {
      const cachedCourses = getCachedData<Course[]>('allCourses');
      if (cachedCourses) {
        setCourses(cachedCourses);
      } else {
        const allCourses = await getAllCourses();
        setCourses(allCourses);
        setCachedData('allCourses', allCourses);
      }

      if (authUser) {
        const cachedUserCourses = getCachedData<Course[]>(`userCourses_${authUser.uid}`);
        if (cachedUserCourses) {
          setUserCourses(cachedUserCourses);
        } else {
          const userCoursesData = await getUserCoursesAPI(authUser.uid);
          setUserCourses(userCoursesData);
          setCachedData(`userCourses_${authUser.uid}`, userCoursesData);
        }
      } else {
        setUserCourses([]);
      }
    } catch (error) {
      console.error('Error fetching courses:', error);
    } finally {
      setLoading(false);
    }
  }, [authUser]);

  useEffect(() => {
    fetchCourses();
  }, [fetchCourses]);

  const addCourse = async (courseId: string) => {
    if (!authUser) return;
    try {
      await addCourseToUser(authUser.uid, courseId);
      const updatedUserCourses = await getUserCoursesAPI(authUser.uid);
      setUserCourses(updatedUserCourses);
    } catch (error) {
      console.error('Error adding course:', error);
      throw error;
    }
  };

  const getCourse = async (courseId: string): Promise<Course | null> => {
    try {
      return await getCourseById(courseId);
    } catch (error) {
      console.error('Error fetching course:', error);
      return null;
    }
  };

  const getWorkout = useCallback(async (workoutId: string): Promise<Workout | null> => {
    try {
      return await getWorkoutById(workoutId);
    } catch (error) {
      console.error('Error fetching workout:', error);
      return null;
    }
  }, []);

  const getUserProfile = async (): Promise<User | null> => {
    if (!authUser) return null;
    try {
      return await getUserProfileFromAPI(authUser.uid);
    } catch (error) {
      console.error('Error fetching user profile:', error);
      return null;
    }
  };

  const updateProfile = async (data: Partial<User>): Promise<void> => {
    if (!authUser) return;
    try {
      await updateUserProfile(authUser.uid, data);
    } catch (error) {
      console.error('Error updating user profile:', error);
      throw error;
    }
  };
  
  const getUserCourses = useCallback(async (userId: string): Promise<Course[]> => {
    try {
      return await getUserCoursesAPI(userId);
    } catch (error) {
      console.error('Error fetching user courses:', error);
      return [];
    }
  }, []);

  const resetCourseProgress = async (userId: string, courseId: string) => {
    if (!userId || !courseId) return;
    try {
      const courseRef = ref(database, `courses/courses/${courseId}`);
      const courseSnapshot = await get(courseRef);
      if (courseSnapshot.exists()) {
        const courseData = courseSnapshot.val();
        const workoutIds = courseData.workouts || [];
        const progressRef = ref(database, `userProgress/${userId}`);
        const progressSnapshot = await get(progressRef);
        if (progressSnapshot.exists()) {
          const progressData = progressSnapshot.val();
          for (const workoutId of workoutIds) {
            if (progressData[workoutId]) {
              await remove(ref(database, `userProgress/${userId}/${workoutId}`));
            }
          }
        }
      }
    } catch (error) {
      console.error('Error resetting course progress:', error);
    }
  };

  return {
    courses,
    userCourses,
    loading,
    addCourse,
    getCourse,
    getWorkout,
    getUserProfile,
    updateProfile,
    getUserCourses,
    resetCourseProgress,
  };
};