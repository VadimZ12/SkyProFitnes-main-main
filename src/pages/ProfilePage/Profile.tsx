import React from 'react'
import Modal from '../../components/Modals/Modal'
import { useAuth } from '../../hooks/useAuth'

function Profile() {
  const { user, logout, updateUser } = useAuth()
  const [isModalOpen, setIsModalOpen] = React.useState(false)
  const [modalType, setModalType] = React.useState<'newPassword'>('newPassword')
  const [isEditing, setIsEditing] = React.useState(false)
  const [editedName, setEditedName] = React.useState(user?.customDisplayName || user?.displayName || user?.email?.split('@')[0] || '')
  const [saveMessage, setSaveMessage] = React.useState('')

  const handleChangePassword = () => {
    setModalType('newPassword')
    setIsModalOpen(true)
  }

	const handleLogout = async () => {
		try {
			await logout()
			window.location.href = '/'
		} catch (error) {
			console.error('Error logging out:', error)
		}
	}

  const handleSaveName = async () => {
    if (user) {
      try {
        await updateUser({ displayName: editedName });
        setIsEditing(false);
        setSaveMessage('Имя успешно сохранено');
        setTimeout(() => setSaveMessage(''), 3000);
      } catch (error) {
        console.error('Error updating display name:', error);
        setSaveMessage('Ошибка при сохранении имени');
        setTimeout(() => setSaveMessage(''), 3000);
      }
    }
  };

  return (
    <section className='flex mobile:flex-col mobile:items-center gap-[33px] w-[100%] p-[30px] mt-[40px] mb-[60px] rounded-[30px] shadow-[0px_4px_67px_-12px_rgba(0,0,0,0.13)]'>
      <img src={user?.photoURL || '/images/profile_no-img.png'} className='w-[197px] h-[197px] rounded-[30px] mobile:w-[141px] mobile:h-[141px]' loading="lazy"/>
      <div className='flex flex-col gap-[30px] mobile:w-full'>
        {isEditing ? (
          <div className='flex items-center'>
            <input
              type="text"
              value={editedName}
              onChange={(e) => setEditedName(e.target.value)}
              className='text-left font-roboto text-[32px] font-medium leading-[35.2px] border-b-2 border-black focus:outline-none'
            />
            <button onClick={handleSaveName} className='ml-2 text-green'>
              Сохранить
            </button>
          </div>
        ) : (
          <h2 className='text-left font-roboto text-[32px] font-medium leading-[35.2px] flex items-center'>
            {user?.displayName || user?.email?.split('@')[0]}
            <button onClick={() => setIsEditing(true)} className='ml-2 text-gray'>
              ✎
            </button>
          </h2>
        )}
        {saveMessage && <p className="text-green">{saveMessage}</p>}
        <div className='flex flex-col gap-[10px] text-left font-roboto text-[18px] font-normal leading-[19.8px]'>
          <p>Логин: {user?.email || user?.login}</p>
          <p>Пароль: ********</p>
        </div>
        <div className='flex flex-wrap gap-[10px] mobile:flex-col'>
          <button onClick={handleChangePassword} className='w-[192px] mobile:w-full h-[52px] rounded-btnRad bg-green'>
            <p className='text-black font-roboto text-[18px] font-normal leading-[19.8px]'>
              Изменить пароль
            </p>
          </button>
          <button onClick={handleLogout} className='w-[192px] mobile:w-full h-[52px] rounded-btnRad bg-white border border-black'>
            <p className='text-black font-roboto text-[18px] font-normal leading-[19.8px]'>
              Выйти
            </p>
          </button>
        </div>
        </div>
      <Modal isOpen={isModalOpen} onClose={() => setIsModalOpen(false)} type={modalType} onSwitchType={() => {}} />
    </section>
  )
}

export default Profile