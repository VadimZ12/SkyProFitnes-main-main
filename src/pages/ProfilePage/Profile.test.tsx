import React from 'react'
import { render, fireEvent, waitFor } from '@testing-library/react'
import Profile from './Profile'
import { useAuth } from '../../hooks/useAuth'
import { MemoryRouter } from 'react-router-dom'
import '@testing-library/jest-dom'

jest.mock('../../hooks/useAuth', () => ({
	useAuth: jest.fn(),
}))

jest.mock('../../components/Modals/Modal', () =>
	jest.fn(() => <div role='dialog' />)
)
jest.mock('../../config/firebase', () => ({
	auth: { signInWithEmailAndPassword: jest.fn(), signOut: jest.fn() },
	database: { ref: jest.fn(), get: jest.fn(), set: jest.fn() },
}))

describe('Profile component', () => {
	beforeAll(() => {
		// eslint-disable-next-line @typescript-eslint/no-explicit-any
		;(global as any).importMeta = {
			env: {
				VITE_FIREBASE_API_KEY: 'mock-api-key',
				VITE_FIREBASE_AUTH_DOMAIN: 'mock-auth-domain',
				VITE_FIREBASE_DATABASE_URL: 'mock-database-url',
				VITE_FIREBASE_PROJECT_ID: 'mock-project-id',
				VITE_FIREBASE_STORAGE_BUCKET: 'mock-storage-bucket',
				VITE_FIREBASE_MESSAGING_SENDER_ID: 'mock-messaging-sender-id',
				VITE_FIREBASE_APP_ID: 'mock-app-id',
			},
		}
	})

	it('renders correctly when the user is logged in', () => {
		;(useAuth as jest.Mock).mockReturnValue({
			user: {
				displayName: 'John Doe',
				email: 'johndoe@example.com',
				photoURL: '/images/profile.png',
			},
			logout: jest.fn(),
			updateUser: jest.fn(),
		})

		const { asFragment, getByText } = render(
			<MemoryRouter>
				<React.Fragment>
					<Profile />
				</React.Fragment>
			</MemoryRouter>
		)

		expect(getByText('John Doe')).toBeInTheDocument()
		expect(asFragment()).toMatchSnapshot()
	})

	it('opens edit mode when edit button is clicked', () => {
		;(useAuth as jest.Mock).mockReturnValue({
			user: {
				displayName: 'John Doe',
				email: 'johndoe@example.com',
				photoURL: '/images/profile.png',
			},
			logout: jest.fn(),
			updateUser: jest.fn(),
		})

		const { getByText, getByDisplayValue } = render(
			<MemoryRouter>
				<React.Fragment>
					<Profile />
				</React.Fragment>
			</MemoryRouter>
		)

		fireEvent.click(getByText('✎'))

		expect(getByDisplayValue('John Doe')).toBeInTheDocument()
		expect(getByText('Сохранить')).toBeInTheDocument()
	})

	it('opens the modal when the change password button is clicked', () => {
		;(useAuth as jest.Mock).mockReturnValue({
			user: {
				displayName: 'John Doe',
				email: 'johndoe@example.com',
				photoURL: '/images/profile.png',
			},
			logout: jest.fn(),
			updateUser: jest.fn(),
		})

		const { getByText, getByRole } = render(
			<MemoryRouter>
				<React.Fragment>
					<Profile />
				</React.Fragment>
			</MemoryRouter>
		)

		fireEvent.click(getByText('Изменить пароль'))

		expect(getByRole('dialog')).toBeInTheDocument()
	})

	it('logs out the user when the logout button is clicked', async () => {
		const logoutMock = jest.fn()

		;(useAuth as jest.Mock).mockReturnValue({
			user: {
				displayName: 'John Doe',
				email: 'johndoe@example.com',
			},
			logout: logoutMock,
			updateUser: jest.fn(),
		})

		const { getByText } = render(
			<MemoryRouter>
				<React.Fragment>
					<Profile />
				</React.Fragment>
			</MemoryRouter>
		)

		fireEvent.click(getByText('Выйти'))

		await waitFor(() => {
			expect(logoutMock).toHaveBeenCalled()
		})
	})
})
