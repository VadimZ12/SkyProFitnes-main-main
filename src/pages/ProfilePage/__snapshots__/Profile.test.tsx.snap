// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile component renders correctly when the user is logged in 1`] = `
<DocumentFragment>
  <section
    class="flex mobile:flex-col mobile:items-center gap-[33px] w-[100%] p-[30px] mt-[40px] mb-[60px] rounded-[30px] shadow-[0px_4px_67px_-12px_rgba(0,0,0,0.13)]"
  >
    <img
      class="w-[197px] h-[197px] rounded-[30px] mobile:w-[141px] mobile:h-[141px]"
      loading="lazy"
      src="/images/profile.png"
    />
    <div
      class="flex flex-col gap-[30px] mobile:w-full"
    >
      <h2
        class="text-left font-roboto text-[32px] font-medium leading-[35.2px] flex items-center"
      >
        John Doe
        <button
          class="ml-2 text-gray"
        >
          ✎
        </button>
      </h2>
      <div
        class="flex flex-col gap-[10px] text-left font-roboto text-[18px] font-normal leading-[19.8px]"
      >
        <p>
          Логин: johndoe@example.com
        </p>
        <p>
          Пароль: ********
        </p>
      </div>
      <div
        class="flex flex-wrap gap-[10px] mobile:flex-col"
      >
        <button
          class="w-[192px] mobile:w-full h-[52px] rounded-btnRad bg-green"
        >
          <p
            class="text-black font-roboto text-[18px] font-normal leading-[19.8px]"
          >
            Изменить пароль
          </p>
        </button>
        <button
          class="w-[192px] mobile:w-full h-[52px] rounded-btnRad bg-white border border-black"
        >
          <p
            class="text-black font-roboto text-[18px] font-normal leading-[19.8px]"
          >
            Выйти
          </p>
        </button>
      </div>
    </div>
    <div
      role="dialog"
    />
  </section>
</DocumentFragment>
`;
